src/omp_autotune.c

```diff
@@ -0,0 +1,120 @@
+/*
+ * This software is Copyright (c) 2017 magnum
+ * and is hereby released to the general public under the following terms:
+ * Redistribution and use in source and binary forms, with or without
+ * modifications, are permitted.
+ */
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+#include "john.h"
+#include "options.h"
+#include "logger.h"
+#include "formats.h"
+#include "timer.h"
+#include "memdbg.h"
+
+extern volatile int bench_running;
+
+int omp_autotune(struct fmt_main *self)
+{
+	static int omp_autotune_running;
+	int threads = omp_get_max_threads();
+	int mkpc = self->params.max_keys_per_crypt * threads;
+	int best_scale = 1, scale = 1;
+	int best_cps = 0;
+	int no_progress = 0;
+	char key[] = "testkey0";
+	sTimer timer;
+	double duration;
+
+	if (omp_autotune_running || threads == 1)
+		return threads;
+
+	omp_autotune_running = 1;
+
+	self->params.min_keys_per_crypt *= threads;
+
+	self->methods.set_salt(
+		self->methods.salt(self->params.tests[0].ciphertext));
+
+	sTimer_Init(&timer);
+
+	if (john_main_process &&
+	    options.verbosity > VERB_DEFAULT && bench_running)
+		fprintf(stderr, "\n");
+
+	do {
+		int i;
+		int this_kpc = mkpc * scale;
+		int cps, crypts = 0;
+
+		self->params.max_keys_per_crypt = this_kpc;
+
+		// Set up buffers
+		self->methods.init(self);
+
+		// Load keys
+		self->methods.clear_keys();
+		for (i = 0; i < this_kpc; i++) {
+			key[7] = '0' + i % 10;
+			self->methods.set_key(key, i);
+		}
+
+		// Warm-up run
+		{
+			int count = this_kpc;
+
+			self->methods.crypt_all(&count, NULL);
+		}
+
+		sTimer_Start(&timer, 1);
+		do {
+			int count = this_kpc;
+
+			self->methods.crypt_all(&count, NULL);
+			crypts += count;
+		} while (sTimer_GetSecs(&timer) < 0.010);
+		sTimer_Stop(&timer);
+
+		// Release buffers
+		self->methods.done();
+
+		duration = sTimer_GetSecs(&timer);
+		cps = crypts / duration;
+
+		if (john_main_process && options.verbosity == VERB_MAX)
+			fprintf(stderr, "scale %d: %d crypts in %f seconds, %d c/s",
+			        scale, crypts, duration, (int)(crypts / duration));
+
+		// Example, require 5% boost (because a low KPC has some advantages)
+		if (cps >= (best_cps * 1.05)) {
+			if (john_main_process && options.verbosity == VERB_MAX)
+				fprintf(stderr, " +\n");
+			best_cps = cps;
+			best_scale = scale;
+		}
+		else {
+			if (john_main_process && options.verbosity == VERB_MAX)
+				fprintf(stderr, " -\n");
+			no_progress++;
+		}
+
+		// Double each time
+		scale *= 2;
```

Comment:
how about:
```c
scale += (scale)/2+1;
```
That would chose for values under 200k : 
1 2 4 7 11 17 26 40 61 92 139 209 314 472 709 1064 1597 2396 3595 5393 8090 12136 18205 27308 40963 61445 92168 138253
vs:
1 2 4 8 16 32 64 128 256 512 1024 2048 4096 8192 16384 32768 65536 131072

Just gives a bit more granularity.

Or possibly this expression:

```c
scale += scale < 5 ? 1 ? (scale/3)+1;
```
which gives this result (even better granularity):
1 2 3 4 5 8 12 18 26 36 50 68 92 124 167 224 300 402 538 719 960 1282 1711 2283 3046 4063 5419 7227 9638 12852 17138 22852 30471 40630 54175 72235 96315 128422 171231

Reply:
With a very limited testing (of the formats you already have done), the scale *=2 seems to work as good or better than either of the other 2 'finer' granular methods I proposed, even when I messed around with the ```if (cps >= (best_cps * 1.05))``` code a bit.

KISS method looks damn good ;)